File: leet-code/121. Best Time to Buy and Sell Stock/BestTimeToBuyAndSellStock121_3.java
package com.example;

public class BestTimeToBuyAndSellStock121_3 {
    public int maxProfit(int[] prices) {
        // 최소값 보다 작은 값이 나오면 최소값 갱신
        // 최소값 보다 크거나 같은 값이 나오면 최소값과 빼서 최대값 비교
        int minStock = 10000;
        int maxProfit = 0;
        for (int price : prices) {
            if (minStock > price) {
                minStock = price;
            } else {
                maxProfit = Math.max(maxProfit, price - minStock);
            }
        }
        return maxProfit;
    }
}

File: leet-code/121. Best Time to Buy and Sell Stock/BestTimeToBuyAndSellStock121_3.kt
package com.example

class BestTimeToBuyAndSellStock121_3 {
    fun maxProfit(prices: IntArray): Int {
        var profit = 0
        var prev = 10001
        for (price in prices) {
            prev = prev.coerceAtMost(price)
            profit = profit.coerceAtLeast(price - prev)
        }
        return profit
    }
}
